Feat/websockets connect

ketrab321 opened the pull request:
#14 
#13

ketrab321 committed:
FastApi Websocket test app

ketrab321 committed:
Sending servers html

ketrab321 committed:
Server prototype

ketrab321 committed:
Server prototype

ketrab321 committed:
Changes in __init__.py and __main__.py

ketrab321 committed:
Websocket chat test

ketrab321 committed:
Taking table id from url

ketrab321 committed:
Saving usernames, and upgraded chat to recognize usernames

ketrab321 committed:
Merge branch 'master' into feat/websockets-connect

ketrab321 committed:
Created connection router and establishing connections with server via websockets

ketrab321 committed:
Changed User to Player

ketrab321 committed:
Routers change

KyrietS committed:
Optimised imports + code refactor

KyrietS committed:
Reformats

KyrietS reviewed (changes requested):
Jak klient zamknie kartę przeglądarki, to rzucany jest wyjątek i nie jest nigdzie łapany. Serwer po takim błędzie się psuje i nowi gracze nie mogą się już połączyć. Trzeba dodać jakąś minimalną obsługę zakończenia połączenia.

KyrietS commented on blankiety_galantow/routers.py:
```diff
@@ -1,10 +1,39 @@
+    else:
+        app.server.tables[game_id] = [Player(username, websocket)]
+    while True:
+        data = await websocket.receive_text()
```
Tutaj jest rzucany wyjątek jak gracz się rozłączy.

kichel98 commented on blankiety_galantow/app.py:
```diff
@@ -5,15 +5,24 @@
+
 os.chdir(os.path.dirname(__file__))
 
+# Creating FastApi instance and mounting static files
```
Dzięki za komentarze, których ja nie zrobiłem 👍 Biję się w pierś

KyrietS replied:
To Bartek, ale i tak należy się zastanowić nad sensem komentarzy typu "Creating server object" nad linijką `server = Server()`. Trochę się traci czas czytając takie komentarze oczywiste. Szybciej ogarniesz z kodu co się dzieje 😛

KyrietS replied:
Ad. 1. Tak, do usunięcia. Folder o tej nazwie jest niestandardowy i zbyt ogólny (nazwa sugeruje, że wszystkie klasy tam muszą być)

KyrietS committed:
Improved connection managing and code structure

Created core classes for managing connection. Responsibilities have been distributed among different classes.

KyrietS reviewed (approved):
## Wprowadziłem poprawki
* Naprawiłem przechwytywanie wyjątku przy zerwaniu połączenia. Gracz jest teraz kulturalnie usuwany z pokoju, a reszta użytkowników dostaje komunikat serwerowy na czacie: "Gracz xxx opuścił pokój".
* Porozdzielałem odpowiedzialność na kilka klas i funkcji. Wyszła z tego fajna struktura.
* Dodałem klasę `Room`, która nie wie o innych pokojach. Zarządza jedynie "swoimi" graczami.
* Klasa `Server` od teraz zarządza pokojami `Room` i przydziela do nich graczy.
* Klasa `Player` to na razie taka "nakładka" na surowy Socket wyposażona w dwie funkcje umożliwiające wygodną pracę z nim.
* Wiele innych zmian

Zapoznajcie się ze zmianami jeśli macie ochotę i zgłaszajcie sugestie. Dzisiaj ten branch **musi** wejść, bo mamy REST API w kolejce z poprzedniego milestone do złączenia. Kod REST API trzeba przystosować, aby był kompatybilny z obecną strukturą.

kichel98 commented on blankiety_galantow/resources/game/js/main.js:
```diff
@@ -70,26 +74,36 @@ var game = new Vue({
+
+// Recive message from websocket
+socket.onmessage = function(event) {
+    data = JSON.parse(event.data)
```
Lepiej dać `let` albo `const` przed `data`.

KyrietS replied:
```suggestion
    const data = JSON.parse(event.data);
```

kichel98 commented on blankiety_galantow/resources/game/js/main.js:
```diff
@@ -1,3 +1,5 @@
+socket = connect()
```
```suggestion
const socket = connect()
```
lub
```suggestion
let socket = connect()
```

kichel98 commented on blankiety_galantow/resources/game/js/websocket_connect.js:
```diff
@@ -0,0 +1,26 @@
+function getCookie(cname) {
+    var name = cname + "=";
+    var decodedCookie = decodeURIComponent(document.cookie);
+    var ca = decodedCookie.split(';');
+    for (var i = 0; i < ca.length; i++) {
+        var c = ca[i];
+        while (c.charAt(0) == ' ') {
+            c = c.substring(1);
+        }
+        if (c.indexOf(name) == 0) {
+            return c.substring(name.length, c.length);
+        }
+    }
+    return "";
+}
+
+
+function connect() {
+    parts = document.location.href.split('/')
+    game_id = parseInt(parts[parts.length - 1])
+    console.log(game_id)
+    username = getCookie("username")
+    socket = new WebSocket("ws://localhost:80/connect/" + game_id + "?username=" + username)
+    console.log("XDDD")
+    return socket
+}
```
Warto wszystkie `var` zamienić na `let` lub `const`, analogicznie, gdy w ogóle nie ma explicite napisanego `var` (jest domyślnie).

KyrietS replied:
Tutaj w ogóle użycie *Cookie* jest zupełnie niepotrzebne. Po co wysyłać to ciasteczko do serwera za każdym razem? Trzeba to przerobić na prosty [LocalStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) klucz-wartość, bez tego dziwnego parsowania ciasteczka 😄 
To co się dzieje w tej funkcji getCookie, to trochę magia i będzie z tego task ;D

`connect` zaraz poprawię

KyrietS replied:
Issue #20 jest do przerobienia funkcji `getCookie`. Tymczasowo można chyba zatwierdzić tę brzydką wersję

kichel98 commented on blankiety_galantow/core/player.py:
```diff
@@ -0,0 +1,26 @@
+        """Send JSON data to the player via socket."""
+        await self.socket.send_json(data)
+
+    name: str
+    socket: WebSocket
```
1. Potrzebujemy tych dwóch linijek? Co dodają?
2. Jeśli mają zostać, to może warto je dać nad konstruktor, jeśli to są pola klasy.

KyrietS replied:
Właśnie też nie do końca rozumiem tych linii. One chyba są po do, żeby przy korzystaniu z obiektu `player.socket` podpowiadało, że pole `socket` jest typu `WebSocket`. Też bym wolał, aby to było w konstruktorze gdzieś wrzucone, ale nie sprawdzałem czy się da...

KyrietS replied:
Zrobiłem testy i te dodatkowe linijki są zbędne. PyCharm przynajmniej potrafi ogarnąć typ z samego konstruktora, w którym są oznaczone typy.

KyrietS replied:
Usunięte w 2765026

kichel98 commented on blankiety_galantow/core/room.py:
```diff
@@ -0,0 +1,58 @@
+from typing import Dict
```
Fajne to! 😄 Byłoby spoko, gdybyśmy wszyscy używali, ale pewnie będzie ciężko 😉

KyrietS replied:
Można próbować stosować typy zmiennych gdzie się da. Zobaczymy jak to wyjdzie 😉

kichel98 commented on blankiety_galantow/core/room.py:
```diff
@@ -0,0 +1,58 @@
+from typing import Dict
+from starlette.websockets import WebSocketDisconnect
```
Jeśli się da, to można zamienić `starlette` na `fastapi`, ale to pierdoooła

KyrietS replied:
Szukałem długo jak zaimportować ten nieszczęsny `WebSocketDisconnect`. Nigdzie nie ma dokumentacji co to jest wgl xD
W końcu znalazłem w issue w repo FastAPI ktoś wysłał przykładowy kod i właśnie tak zaimportował to. Można to poprawić 👍 
```suggestion
from fastapi.websockets import WebSocketDisconnect
```

kichel98 commented on blankiety_galantow/core/server.py:
```diff
@@ -0,0 +1,24 @@
+        """Add new player to existing room."""
+        await self._rooms[room_id].add_player_and_listen(player)
+
+    _rooms: Dict[int, Room]
```
Takie samo pytanie, jak w player.py, ostatnia linia jest tylko do zaznaczenia typu zmiennej `_rooms`?

KyrietS replied:
Usunięte w 2765026

kichel98 commented on blankiety_galantow/resources/js/main.js:
```diff
@@ -1,5 +1,19 @@
+function setCookie(cname, cvalue, exdays) {
+    var d = new Date();
+    d.setTime(d.getTime() + (exdays * 24 * 60 * 60 * 1000));
+    var expires = "expires=" + d.toUTCString();
+    document.cookie = cname + "=" + cvalue + ";" + expires + ";path=/";
+}
```
Tak samo, `var` -> `let` lub `const`

KyrietS replied:
Issue #20. Tymczasowo zostawiamy

kichel98 commented on blankiety_galantow/routers.py:
```diff
@@ -1,10 +1,35 @@
+@router.get("/game/{game_id}")
+async def connect_to_game(game_id):
+    f = codecs.open('resources/game/index.html', 'r', 'utf-8')
+    return HTMLResponse(f.read())
```
Te dwie funkcje robią tak naprawdę dokładnie to samo, a jedyna różnica, to ścieżka do HTMLa. Byłoby spoko, jakby wyglądały tak samo (obojętnie czy tak jak pierwsza, czy tak jak druga, oba sposoby są spoko). Można też ewentualnie zrobić osobną funkcję w stylu `read_html`.

KyrietS replied:
Rozwiązane w commitcie 9021f9d.

kichel98 commented on blankiety_galantow/resources/game/js/main.js:
```diff
@@ -70,26 +79,38 @@ var game = new Vue({
+});
+
+
+// Recive message from websocket
```
```suggestion
// Receive message from websocket
```

kichel98 reviewed (changes requested):
1. Mam pusty folder "classes", on jest do usunięcia?
2. Dobra robota, fajnie to zaczyna wyglądać 👍

kichel98 commented on blankiety_galantow/resources/game/js/websocket_connect.js:
```diff
@@ -0,0 +1,26 @@
+    console.log(game_id)
+    username = getCookie("username")
+    socket = new WebSocket("ws://localhost:80/connect/" + game_id + "?username=" + username)
+    console.log("XDDD")
```
XDDD

KyrietS replied:
XDDD
```suggestion
```

KyrietS committed:
Added 'let' specifier for variable

Co-authored-by: kichel98 <44356453+kichel98@users.noreply.github.com>

KyrietS committed:
Missing semicolon

KyrietS committed:
Added 'const' specifier

KyrietS committed:
Spelling

Co-authored-by: kichel98 <44356453+kichel98@users.noreply.github.com>

KyrietS committed:
Removed redundant log message

KyrietS committed:
Changed import source from starlette to fastapi

KyrietS committed:
Code refactor

KyrietS committed:
Refactor a function so it uses the same convention

KyrietS committed:
Removed redundant type-markers
